src/aspire/storage/starfile.py

```diff
@@ -1,173 +1,181 @@
+                continue
+            if len(df) == 1:
+                for col in list(df):
+                    # key is the column label. value is value in the df at row 0, column col
+                    _block.set_pair(col, str(df.at[0, col]))
+            elif len(df) > 1:
+                # initialize loop with column names
+                _loop = _block.init_loop("", list(df.columns))
+                for row in df.values.tolist():
+                    row = [str(row[x]) for x in range(len(row))]
+                    _loop.add_row(row)
+
+        _doc.write_file(filepath)
+
+    def get_block_by_index(self, index):
+        """
+        Retrieve a DataFrame representing a star file block by its position in the starfile
+        """
+        return self.blocks[list(self.blocks.keys())[index]]
 
-        logger.info("Initializing StarFile object from data")
-        self.init_from_blocks(blocks)
-        logger.info(f"Created <{self}>")
+    def __getitem__(self, key):
+        """
+        Retrieve the star file block with name `key`
+        """
+        return self.blocks[key]
 
-    def init_from_blocks(self, blocks):
+    def __setitem__(self, key, value):
         """
-        Initialize a StarFile from a list of blocks
-        :param blocks: An iterable of StarFileBlock objects
-        :return: An initialized StarFile object
+        Pass in a Pandas Dataframe to add a block named `key` with values `value`
         """
-        for block in blocks:
-            self.blocks[block.name] = block
+        self.blocks[key] = value
 
     def __repr__(self):
-        return f"StarFile with {len(self.blocks)} blocks"
+        return "StarFile with blocks: " + ", ".join(self.blocks.keys())
 
-    def __getitem__(self, item):
-        if isinstance(item, str):
-            return self.blocks[item]
-        else:
-            return self.blocks[list(self.blocks.keys())[item]]
+    def __iter__(self):
+        return self.blocks.items().__iter__()
 
     def __len__(self):
         return len(self.blocks)
 
     def __eq__(self, other):
-        return all(b1 == b2 for b1, b2 in zip(self.blocks, other.blocks))
-
-    def save(self, f):
-        for block in self:
-            f.write(f"data_{block.name}\n\n")
-            if block.properties is not None:
-                for k, v in block.properties.items():
-                    f.write(f"{k} {v}\n")
-            f.write("\n")
-            for loop in block:
-                f.write("loop_\n")
-                for col in loop.columns:
-                    f.write(f"{col}\n")
-                for _, row in loop.iterrows():
-                    f.write(" ".join(map(str, row)) + "\n")
-                f.write("\n")
+        if not len(self) == len(other):
+            return False
+        self_list = list(self.blocks.keys())
+        other_list = list(other.blocks.keys())
+        for i in range(len(self_list)):
+            if not self_list[i][0] == other_list[i][0]:
```

Comment:
What's this doing? Why do we only need to test [0] and [1]?

Reply:
Each element of `self_list` and `other_list` is a tuple with `(key, value)`. So I'm first testing the key and then testing the DF it indexes. I saw a mistake here, where the second line should use `DataFrame.equals()`, not `==`. Fixed.

Reply:
The reason I can't use `OrderedDict.__eq__()` here is that `__eq__` isn't defined for DataFrame

Reply:
Hence breaking into the list and comparing item by item

Reply:
This should also be `items()` not `keys()`. Thank you for catching

Reply:
This was fixed, and on the way I caught another bug that happens when initializing the StarFile from blocks. There is now a line in the constructor that converts given DataFrame to `dtype=str`. This way, the equality check will never fail due to differing types